### PR TITLE
Universal support for DimensionsHelper and  ScrollbarHelper

### PR DESCRIPTION
**Pull request by earlyster**
**What kind of change does this PR introduce?** (check one with "x")
- [x ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

Error when you load ngx-datagrid scrollbar width is calculated and grid width is calculated based on window size. With universal SSR we can't calculate it that same way so we need to replace implementation of DimensionHelper with a universal helpful version

**What is the new behavior?**

No error and you must implement your own DimensionsHelper. Here is a snippet of my version of SSR DimensionsHelper. I had to export this service in order to inject a different implementation. 

```
import { ScrollbarHelper } from '@swimlane/ngx-datatable';
import { Injectable, Inject } from '@angular/core';
import { Request } from 'express';
import { REQUEST } from '@nguniversal/express-engine/tokens';
import { DimensionsHelper } from '@swimlane/ngx-datatable';

@Injectable()
export class ServerDimensionsHelper extends DimensionsHelper {

    constructor(@Inject(REQUEST) private request: Request) {
        super();
    }


    getDimensions(element: Element): ClientRect {
        const width = parseInt(this.request.cookies['CH-DW'], 10) || 1000;
        const height = parseInt(this.request.cookies['CH-DH'], 10) || 800;

        const adjustedWidth = width;
        const adjustedHeight = height;


        return {
            height: adjustedHeight,
            bottom: 0,
            top: 0,
            width: adjustedWidth,
            left: 0,
            right: 0
        };
    }
}
```

Also had to create a ScrollbarHelper service implementation for server. Which returns my custom css style width of all scrollbars of 16.

```
import { ScrollbarHelper } from '@swimlane/ngx-datatable';
import { Injectable, Inject } from '@angular/core';
import { DOCUMENT } from '@angular/common';

@Injectable()
export class ServerScrollBarHelper extends ScrollbarHelper {
    width: number;

    constructor(@Inject(DOCUMENT) document) {
        super(document);
        this.width = 16;
    }

    getWidth(): number {
        return this.width;
    }
}

```

I then added these two services using @NgModule providers

```
  providers:  [  {
    provide: ScrollbarHelper,
    useClass: ServerScrollBarHelper
  },
  {
    provide: DimensionsHelper,
    useClass: ServerDimensionsHelper
  }]
```

**Does this PR introduce a breaking change?** (check one with "x")
- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

I did not include these sample implementations in my pull request since this strategy may not work for everyone.

### CODE REVIEW

**Comment by amcdnl** [comment]
Thanks for the PR, this looks awesome. Can you remove the built files?

File: release/components/body/scroller.component.d.ts [hunk 1]
```diff
@@ -13,7 +14,7 @@ export declare class ScrollerComponent implements OnInit, OnDestroy {
     element: any;
     parentElement: any;
     onScrollListener: any;
-    constructor(ngZone: NgZone, element: ElementRef);
+    constructor(ngZone: NgZone, element: ElementRef, renderer: Renderer2);
```
Comment (amcdnl) [inline release/components/body/scroller.component.d.ts]: Why did you add Renderer here?

File: src/components/body/scroller.component.ts [hunk 2]
```diff
@@ -36,17 +36,21 @@ export class ScrollerComponent implements OnInit, OnDestroy {
...
   ngOnInit(): void {
     // manual bind so we don't always listen
     if (this.scrollbarV || this.scrollbarH) {
-      this.parentElement = this.element.parentElement.parentElement;
+      const renderer = this.renderer;
+      this.parentElement = renderer.parentNode(renderer.parentNode(this.element));
       this.ngZone.runOutsideAngular(() => {
-        this.parentElement.addEventListener('scroll', this.onScrolled.bind(this));
+      renderer.listen(
+        this.parentElement, 'scroll', this.onScrolled.bind(this));
```
Comment (amcdnl) [inline src/components/body/scroller.component.ts]: You actually shouldn't have to use the renderer to do this in v5.

File: src/directives/resizeable.directive.ts [hunk 3]
```diff
@@ -24,15 +24,17 @@ export class ResizeableDirective implements OnDestroy, AfterViewInit {
...
   }
 
   ngAfterViewInit(): void {
+    const renderer2 = this.renderer;
     if (this.resizeEnabled) {
-      const node = document.createElement('span');
-      node.classList.add('resize-handle');
-      this.element.appendChild(node);
+      const node = renderer2.createElement('span');
+      // node.classList.add('resize-handle');
```
Comment (amcdnl) [inline src/directives/resizeable.directive.ts]: Please remove dead code.

File: src/services/dimensions-helper.service.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+/**
+ * Gets the width of the scrollbar.  Nesc for windows
+ * http://stackoverflow.com/a/13382873/888165
+ */
+@Injectable()
+export class DimensionsHelper {
+
+  // tslint:disable-next-line:no-empty
+  constructor() { }
```
Comment (amcdnl) [inline src/services/dimensions-helper.service.ts]: Why empty constructor?

File: src/components/datatable.component.ts [hunk 5]
```diff
@@ -808,7 +813,8 @@ export class DatatableComponent implements OnInit, DoCheck, AfterViewInit {
    *
    */
   recalculateDims(): void {
-    const dims = this.element.getBoundingClientRect();
+    // const dims = this.element.getBoundingClientRect();
```
Comment (amcdnl) [inline src/components/datatable.component.ts]: Please remove dead code.

File: src/components/body/scroller.component.ts (hunk 2 above)
Comment (earlyster) [reply src/components/body/scroller.component.ts]: I originally started with v4

File: src/components/datatable.component.ts (hunk 5 above)
Comment (earlyster) [reply src/components/datatable.component.ts]: Done

**Comment by earlyster** [comment]
PR should have all clean up now

**Comment by amcdnl** [comment]
One last request, can you add your example config to the docs?

**Comment by earlyster** [comment]
Added universal doc

**Comment by amcdnl** [comment]
Thanks for your contribution!!!

**Comment by earlyster** [comment]
Your welcome
